### ISSUE TITLE
Extensions endpoint returns certain item_data as objects or as arrays - Fatal errors on PHP >=8.0

### ISSUE THREAD

**Issue by PanosSynetos** [woocommerce/woocommerce-gutenberg-products-block#7275]
## Describe the bug

After quite some digging, it seems like I found an issue on the StoreAPI, when posting to `/wc/store/v1/cart/extensions` which leads to Fatal Errors on PHP >=8.0

I have gift cards enabled, so that I have a checkbox on cart/checkout to post to `extensions`

## To reproduce

Using the following snippet, and having xdebug enabled, when posting to the `/wc/store/v1/cart/extensions` endpoint, the code gets executed twice.

As you can see in my comments, the first time it runs, certain keys are objects.
The second time, they are an array.

This is causing fatal errors with PHP >=8.0, as when we want to change
`$item_data[ 'totals' ]->line_total` through a plugin, the first time it will pass (object), the second time, it will throw a fatal error

`PHP Fatal error:  Uncaught Error: Attempt to assign property "XXX" on array`

```php
add_filter( 'rest_request_after_callbacks', 'filter_cart_item_data', 10, 3 );
function filter_cart_item_data( $response, $server, $request ) {

	if ( is_wp_error( $response ) ) {
		return $response;
	}
	if ( strpos( $request->get_route(), 'wc/store' ) === false ) {
		return $response;
	}
	$data = $response->get_data();

	if ( empty( $data[ 'items' ] ) ) {
		return $response;
	}

	foreach ( $data[ 'items' ] as &$item_data ) {

		// FIRST TIME 👇
		// $item_data['quantity_limits'] is an object
		// $item_data['prices'] is an object
		// $item_data['totals'] is an object
		// $item_data['extensions'] is an object

		// SECOND TIME 👇
		// $item_data['quantity_limits'] is an array
		// $item_data['prices'] is an array
		// $item_data['totals'] is an array
		// $item_data['extensions'] is an array
	}
	return $response;
}
```

## Expected behavior

Both times, the specific keys, should be objects and not arrays.

### WordPress (please complete the following information):

-   WordPress version: 6.0.2
-   WooCommerce version: 6.9.3

## Additional context

I doubt it’s related to Gift Cards, but I don’t have another plugin in hand that posts to `extensions` to try it out.

**Comment by wavvves** [comment]
@PanosSynetos have you tried logging  $request->get_route()  or even the whole $response  both times? And also $request. Maybe it gets called twice on different endpoints / provides different request parameters.

If that is the case, we must look carefully into this, as the discrepancy might be intentional.

**Comment by PanosSynetos** [comment]
There you go

```

```
🚨🚨🚨🚨🚨🚨 Second time 🚨🚨🚨🚨🚨🚨 
[03-Oct-2022 14:49:43 UTC] ========$request->get_route()========
[03-Oct-2022 14:49:43 UTC] /wc/store/v1/cart
[03-Oct-2022 14:49:43 UTC] ========$response========
[03-Oct-2022 14:49:43 UTC] WP_REST_Response Object
(
    [data] => Array
        (
            [coupons] => Array
                (
                )

            [shipping_rates] => Array
                (
                    [0] => Array
                        (
                            [package_id] => 0
                            [name] => Shipping method
                            [destination] => Array
                                (
                                    [address_1] => 902 Broadwayy
                                    [address_2] =>
                                    [city] => New York
                                    [state] => I
                                    [postcode] => 11143XYZ
                                    [country] => GR
                                )

                            [items] => Array
                                (
                                    [0] => Array
                                        (
                                            [key] => 7f6ffaa6bb0b408017b62254211691b5
                                            [name] => 99 Black Hi
                                            [quantity] => 1
                                        )

                                )

                            [shipping_rates] => Array
                                (
                                    [0] => Array
                                        (
                                            [rate_id] => local_pickup:6
                                            [name] => Local pickup
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 0
                                            [taxes] => 0
                                            [instance_id] => 6
                                            [method_id] => local_pickup
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] => 1
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                    [1] => Array
                                        (
                                            [rate_id] => flat_rate:7
                                            [name] => Flat rate 10
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 1000
                                            [taxes] => 0
                                            [instance_id] => 7
                                            [method_id] => flat_rate
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] =>
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                    [2] => Array
                                        (
                                            [rate_id] => flat_rate:8
                                            [name] => Flat rate 100
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 10000
                                            [taxes] => 0
                                            [instance_id] => 8
                                            [method_id] => flat_rate
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] =>
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                    [3] => Array
                                        (
                                            [rate_id] => free_shipping:9
                                            [name] => Free shipping
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 0
                                            [taxes] => 0
                                            [instance_id] => 9
                                            [method_id] => free_shipping
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] =>
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                )

                        )

                )

            [shipping_address] => Array
                (
                    [first_name] => Panos
                    [last_name] => Test
                    [company] =>
                    [address_1] => 902 Broadwayy
                    [address_2] =>
                    [city] => New York
                    [state] => I
                    [postcode] => 11143XYZ
                    [country] => GR
                    [phone] => 123456779800
                )

            [billing_address] => Array
                (
                    [first_name] => Panos
                    [last_name] => Test
                    [company] =>
                    [address_1] => 902 Broadwayy
                    [address_2] =>
                    [city] => New York
                    [state] => I
                    [postcode] => 11143XYZ
                    [country] => GR
                    [email] => panos@panos.local
                    [phone] => 123456779800
                )

            [items] => Array
                (
                    [0] => Array
                        (
                            [key] => 7f6ffaa6bb0b408017b62254211691b5
                            [id] => 112
                            [quantity] => 1
                            [quantity_limits] => Array
                                (
                                    [minimum] => 1
                                    [maximum] => 9999
                                    [multiple_of] => 1
                                    [editable] => 1
                                )

                            [name] => 99 Black Hi
                            [short_description] => <p>Lightweight and durable. Set of 2. Photo by Martin Talavášek.</p>
                            [description] => <p>Lightweight and durable. Set of 2. Photo by Martin Talavášek.</p>
                            [sku] =>
                            [low_stock_remaining] =>
                            [backorders_allowed] =>
                            [show_backorder_badge] =>
                            [sold_individually] =>
                            [permalink] => https://somewherewarm.local/product/99-black-hi/
                            [images] => Array
                                (
                                    [0] => Array
                                        (
                                            [id] => 144
                                            [src] => https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1.png
                                            [thumbnail] => https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-324x324.png
                                            [srcset] => https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1.png 800w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-324x324.png 324w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-100x100.png 100w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-416x416.png 416w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-300x300.png 300w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-150x150.png 150w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-768x768.png 768w
                                            [sizes] => (max-width: 800px) 100vw, 800px
                                            [name] => truck-99-lo-1.png
                                            [alt] =>
                                        )

                                )

                            [variation] => Array
                                (
                                )

                            [item_data] => Array
                                (
                                )

                            [prices] => Array
                                (
                                    [price] => 1000
                                    [regular_price] => 1000
                                    [sale_price] => 1000
                                    [price_range] =>
                                    [currency_code] => EUR
                                    [currency_symbol] => €
                                    [currency_minor_unit] => 2
                                    [currency_decimal_separator] => ,
                                    [currency_thousand_separator] => .
                                    [currency_prefix] => €
                                    [currency_suffix] =>
                                    [raw_prices] => Array
                                        (
                                            [precision] => 6
                                            [price] => 10000000
                                            [regular_price] => 10000000
                                            [sale_price] => 10000000
                                        )

                                )

                            [totals] => Array
                                (
                                    [line_subtotal] => 1000
                                    [line_subtotal_tax] => 0
                                    [line_total] => 1000
                                    [line_total_tax] => 0
                                    [currency_code] => EUR
                                    [currency_symbol] => €
                                    [currency_minor_unit] => 2
                                    [currency_decimal_separator] => ,
                                    [currency_thousand_separator] => .
                                    [currency_prefix] => €
                                    [currency_suffix] =>
                                )

                            [catalog_visibility] => visible
                            [extensions] => Array
                                (
                                )

                        )

                )

            [items_count] => 1
            [items_weight] => 0
            [cross_sells] => Array
                (
                )

            [needs_payment] => 1
            [needs_shipping] => 1
            [has_calculated_shipping] => 1
            [fees] => Array
                (
                )

            [totals] => Array
                (
                    [total_items] => 1000
                    [total_items_tax] => 0
                    [total_fees] => 0
                    [total_fees_tax] => 0
                    [total_discount] => 0
                    [total_discount_tax] => 0
                    [total_shipping] => 0
                    [total_shipping_tax] => 0
                    [total_price] => 1000
                    [total_tax] => 0
                    [tax_lines] => Array
                        (
                        )

                    [currency_code] => EUR
                    [currency_symbol] => €
                    [currency_minor_unit] => 2
                    [currency_decimal_separator] => ,
                    [currency_thousand_separator] => .
                    [currency_prefix] => €
                    [currency_suffix] =>
                )

            [errors] => Array
                (
                )

            [payment_requirements] => Array
                (
                    [0] => products
                )

            [extensions] => Array
                (
                    [woocommerce-gift-cards] => Array
                        (
                            [balance] => 587005
                            [is_using_balance] =>
                            [pending_total] => 0
                            [cart_total] => 1000
                            [available_total] => 1000
                            [account_giftcards] => Array
                                (
                                )

                            [applied_giftcards] => Array
                                (
                                )

                        )

                )

        )

    [headers] => Array
        (
            [Nonce] => 832c7a90d8
            [Nonce-Timestamp] => 1664808583
            [User-ID] => 1
            [X-WC-Store-API-Nonce] => 832c7a90d8
        )

    [status] => 200
    [links:protected] => Array
        (
        )

    [matched_route:protected] =>
    [matched_handler:protected] =>
)

[03-Oct-2022 14:49:43 UTC] ========$request->get_route()========
[03-Oct-2022 14:49:43 UTC] /wc/store/v1/cart/extensions

```

**Comment by wavvves** [comment]
Also, I just confirmed this is 8.1/8.0 related. Up until 7.4, this would produce a Warning and **won't change the array entry**; from 8.0 onwards, it throws a fatal.

```
<?php
$array = ['something' => '🥑'];
$array->something = "some stuff";
var_dump($array);
```

Results for 8.1.11, 8.0.24:
```
Fatal error: Uncaught Error: Attempt to assign property "something" on array in /home/user/scripts/code.php:5
Stack trace:
#0 {main}
  thrown in /home/user/scripts/code.php on line 5
```
Results for 7.4.32, 7.3.33, 7.2.34:
```
Warning: Attempt to assign property 'something' of non-object in /home/user/scripts/code.php on line 5
array(1) {
  ["something"]=>
  string(4) "🥑"
}

```
It's documented [here](https://www.php.net/manual/en/migration80.incompatible.php) as a breaking change.

**Comment by PanosSynetos** [comment]
There you go

<details><summary>🚨🚨 First Time 🚨🚨 /wc/store/v1/cart</summary>

```
[03-Oct-2022 14:54:28 UTC] WP_REST_Response Object
(
    [data] => Array
        (
            [coupons] => Array
                (
                )

            [shipping_rates] => Array
                (
                    [0] => Array
                        (
                            [package_id] => 0
                            [name] => Shipping method
                            [destination] => stdClass Object
                                (
                                    [address_1] => 902 Broadwayy
                                    [address_2] =>
                                    [city] => New York
                                    [state] => I
                                    [postcode] => 11143XYZ
                                    [country] => GR
                                )

                            [items] => Array
                                (
                                    [0] => Array
                                        (
                                            [key] => 7f6ffaa6bb0b408017b62254211691b5
                                            [name] => 99 Black Hi
                                            [quantity] => 1
                                        )

                                )

                            [shipping_rates] => Array
                                (
                                    [0] => Array
                                        (
                                            [rate_id] => local_pickup:6
                                            [name] => Local pickup
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 0
                                            [taxes] => 0
                                            [instance_id] => 6
                                            [method_id] => local_pickup
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] => 1
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                    [1] => Array
                                        (
                                            [rate_id] => flat_rate:7
                                            [name] => Flat rate 10
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 1000
                                            [taxes] => 0
                                            [instance_id] => 7
                                            [method_id] => flat_rate
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] =>
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                    [2] => Array
                                        (
                                            [rate_id] => flat_rate:8
                                            [name] => Flat rate 100
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 10000
                                            [taxes] => 0
                                            [instance_id] => 8
                                            [method_id] => flat_rate
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] =>
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                    [3] => Array
                                        (
                                            [rate_id] => free_shipping:9
                                            [name] => Free shipping
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 0
                                            [taxes] => 0
                                            [instance_id] => 9
                                            [method_id] => free_shipping
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] =>
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                )

                        )

                )

            [shipping_address] => stdClass Object
                (
                    [first_name] => Panos
                    [last_name] => Test
                    [company] =>
                    [address_1] => 902 Broadwayy
                    [address_2] =>
                    [city] => New York
                    [state] => I
                    [postcode] => 11143XYZ
                    [country] => GR
                    [phone] => 123456779800
                )

            [billing_address] => stdClass Object
                (
                    [first_name] => Panos
                    [last_name] => Test
                    [company] =>
                    [address_1] => 902 Broadwayy
                    [address_2] =>
                    [city] => New York
                    [state] => I
                    [postcode] => 11143XYZ
                    [country] => GR
                    [email] => panos@panos.local
                    [phone] => 123456779800
                )

            [items] => Array
                (
                    [0] => Array
                        (
                            [key] => 7f6ffaa6bb0b408017b62254211691b5
                            [id] => 112
                            [quantity] => 1
                            [quantity_limits] => stdClass Object
                                (
                                    [minimum] => 1
                                    [maximum] => 9999
                                    [multiple_of] => 1
                                    [editable] => 1
                                )

                            [name] => 99 Black Hi
                            [short_description] => <p>Lightweight and durable. Set of 2. Photo by Martin Talavášek.</p>
                            [description] => <p>Lightweight and durable. Set of 2. Photo by Martin Talavášek.</p>
                            [sku] =>
                            [low_stock_remaining] =>
                            [backorders_allowed] =>
                            [show_backorder_badge] =>
                            [sold_individually] =>
                            [permalink] => https://somewherewarm.local/product/99-black-hi/
                            [images] => Array
                                (
                                    [0] => Array
                                        (
                                            [id] => 144
                                            [src] => https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1.png
                                            [thumbnail] => https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-324x324.png
                                            [srcset] => https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1.png 800w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-324x324.png 324w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-100x100.png 100w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-416x416.png 416w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-300x300.png 300w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-150x150.png 150w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-768x768.png 768w
                                            [sizes] => (max-width: 800px) 100vw, 800px
                                            [name] => truck-99-lo-1.png
                                            [alt] =>
                                        )

                                )

                            [variation] => Array
                                (
                                )

                            [item_data] => Array
                                (
                                )

                            [prices] => stdClass Object
                                (
                                    [price] => 1000
                                    [regular_price] => 1000
                                    [sale_price] => 1000
                                    [price_range] =>
                                    [currency_code] => EUR
                                    [currency_symbol] => €
                                    [currency_minor_unit] => 2
                                    [currency_decimal_separator] => ,
                                    [currency_thousand_separator] => .
                                    [currency_prefix] => €
                                    [currency_suffix] =>
                                    [raw_prices] => Array
                                        (
                                            [precision] => 6
                                            [price] => 10000000
                                            [regular_price] => 10000000
                                            [sale_price] => 10000000
                                        )

                                )

                            [totals] => stdClass Object
                                (
                                    [line_subtotal] => 1000
                                    [line_subtotal_tax] => 0
                                    [line_total] => 1000
                                    [line_total_tax] => 0
                                    [currency_code] => EUR
                                    [currency_symbol] => €
                                    [currency_minor_unit] => 2
                                    [currency_decimal_separator] => ,
                                    [currency_thousand_separator] => .
                                    [currency_prefix] => €
                                    [currency_suffix] =>
                                )

                            [catalog_visibility] => visible
                            [extensions] => stdClass Object
                                (
                                )

                        )

                )

            [items_count] => 1
            [items_weight] => 0
            [cross_sells] => Array
                (
                )

            [needs_payment] =>
            [needs_shipping] => 1
            [has_calculated_shipping] => 1
            [fees] => Array
                (
                )

            [totals] => stdClass Object
                (
                    [total_items] => 1000
                    [total_items_tax] => 0
                    [total_fees] => 0
                    [total_fees_tax] => 0
                    [total_discount] => 0
                    [total_discount_tax] => 0
                    [total_shipping] => 0
                    [total_shipping_tax] => 0
                    [total_price] => 0
                    [total_tax] => 0
                    [tax_lines] => Array
                        (
                        )

                    [currency_code] => EUR
                    [currency_symbol] => €
                    [currency_minor_unit] => 2
                    [currency_decimal_separator] => ,
                    [currency_thousand_separator] => .
                    [currency_prefix] => €
                    [currency_suffix] =>
                )

            [errors] => Array
                (
                )

            [payment_requirements] => Array
                (
                    [0] => products
                )

            [extensions] => stdClass Object
                (
                    [woocommerce-gift-cards] => Array
                        (
                            [balance] => 587005
                            [is_using_balance] => 1
                            [pending_total] => 0
                            [cart_total] => 1000
                            [available_total] => 1000
                            [account_giftcards] => Array
                                (
                                    [0] => Array
                                        (
                                            [code] => XXXX-XXXX-XXXX-NYHX
                                            [amount] => 1000
                                        )

                                )

                            [applied_giftcards] => Array
                                (
                                )

                        )

                )

        )

    [headers] => Array
        (
            [Nonce] => 832c7a90d8
            [Nonce-Timestamp] => 1664808868
            [User-ID] => 1
            [X-WC-Store-API-Nonce] => 832c7a90d8
        )

    [status] => 200
    [links:protected] => Array
        (
        )

    [matched_route:protected] =>
    [matched_handler:protected] =>
)
```
</details>



<details><summary>🚨🚨 Second Time 🚨🚨 /wc/store/v1/cart/extensions</summary>

```
[03-Oct-2022 14:54:28 UTC] WP_REST_Response Object
(
    [data] => Array
        (
            [coupons] => Array
                (
                )

            [shipping_rates] => Array
                (
                    [0] => Array
                        (
                            [package_id] => 0
                            [name] => Shipping method
                            [destination] => Array
                                (
                                    [address_1] => 902 Broadwayy
                                    [address_2] =>
                                    [city] => New York
                                    [state] => I
                                    [postcode] => 11143XYZ
                                    [country] => GR
                                )

                            [items] => Array
                                (
                                    [0] => Array
                                        (
                                            [key] => 7f6ffaa6bb0b408017b62254211691b5
                                            [name] => 99 Black Hi
                                            [quantity] => 1
                                        )

                                )

                            [shipping_rates] => Array
                                (
                                    [0] => Array
                                        (
                                            [rate_id] => local_pickup:6
                                            [name] => Local pickup
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 0
                                            [taxes] => 0
                                            [instance_id] => 6
                                            [method_id] => local_pickup
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] => 1
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                    [1] => Array
                                        (
                                            [rate_id] => flat_rate:7
                                            [name] => Flat rate 10
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 1000
                                            [taxes] => 0
                                            [instance_id] => 7
                                            [method_id] => flat_rate
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] =>
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                    [2] => Array
                                        (
                                            [rate_id] => flat_rate:8
                                            [name] => Flat rate 100
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 10000
                                            [taxes] => 0
                                            [instance_id] => 8
                                            [method_id] => flat_rate
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] =>
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                    [3] => Array
                                        (
                                            [rate_id] => free_shipping:9
                                            [name] => Free shipping
                                            [description] =>
                                            [delivery_time] =>
                                            [price] => 0
                                            [taxes] => 0
                                            [instance_id] => 9
                                            [method_id] => free_shipping
                                            [meta_data] => Array
                                                (
                                                    [0] => Array
                                                        (
                                                            [key] => Items
                                                            [value] => 99 Black Hi &times; 1
                                                        )

                                                )

                                            [selected] =>
                                            [currency_code] => EUR
                                            [currency_symbol] => €
                                            [currency_minor_unit] => 2
                                            [currency_decimal_separator] => ,
                                            [currency_thousand_separator] => .
                                            [currency_prefix] => €
                                            [currency_suffix] =>
                                        )

                                )

                        )

                )

            [shipping_address] => Array
                (
                    [first_name] => Panos
                    [last_name] => Test
                    [company] =>
                    [address_1] => 902 Broadwayy
                    [address_2] =>
                    [city] => New York
                    [state] => I
                    [postcode] => 11143XYZ
                    [country] => GR
                    [phone] => 123456779800
                )

            [billing_address] => Array
                (
                    [first_name] => Panos
                    [last_name] => Test
                    [company] =>
                    [address_1] => 902 Broadwayy
                    [address_2] =>
                    [city] => New York
                    [state] => I
                    [postcode] => 11143XYZ
                    [country] => GR
                    [email] => panos@panos.local
                    [phone] => 123456779800
                )

            [items] => Array
                (
                    [0] => Array
                        (
                            [key] => 7f6ffaa6bb0b408017b62254211691b5
                            [id] => 112
                            [quantity] => 1
                            [quantity_limits] => Array
                                (
                                    [minimum] => 1
                                    [maximum] => 9999
                                    [multiple_of] => 1
                                    [editable] => 1
                                )

                            [name] => 99 Black Hi
                            [short_description] => <p>Lightweight and durable. Set of 2. Photo by Martin Talavášek.</p>
                            [description] => <p>Lightweight and durable. Set of 2. Photo by Martin Talavášek.</p>
                            [sku] =>
                            [low_stock_remaining] =>
                            [backorders_allowed] =>
                            [show_backorder_badge] =>
                            [sold_individually] =>
                            [permalink] => https://somewherewarm.local/product/99-black-hi/
                            [images] => Array
                                (
                                    [0] => Array
                                        (
                                            [id] => 144
                                            [src] => https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1.png
                                            [thumbnail] => https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-324x324.png
                                            [srcset] => https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1.png 800w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-324x324.png 324w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-100x100.png 100w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-416x416.png 416w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-300x300.png 300w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-150x150.png 150w, https://somewherewarm.local/wp-content/uploads/2020/11/truck-99-lo-1-768x768.png 768w
                                            [sizes] => (max-width: 800px) 100vw, 800px
                                            [name] => truck-99-lo-1.png
                                            [alt] =>
                                        )

                                )

                            [variation] => Array
                                (
                                )

                            [item_data] => Array
                                (
                                )

                            [prices] => Array
                                (
                                    [price] => 1000
                                    [regular_price] => 1000
                                    [sale_price] => 1000
                                    [price_range] =>
                                    [currency_code] => EUR
                                    [currency_symbol] => €
                                    [currency_minor_unit] => 2
                                    [currency_decimal_separator] => ,
                                    [currency_thousand_separator] => .
                                    [currency_prefix] => €
                                    [currency_suffix] =>
                                    [raw_prices] => Array
                                        (
                                            [precision] => 6
                                            [price] => 10000000
                                            [regular_price] => 10000000
                                            [sale_price] => 10000000
                                        )

                                )

                            [totals] => Array
                                (
                                    [line_subtotal] => 1000
                                    [line_subtotal_tax] => 0
                                    [line_total] => 1000
                                    [line_total_tax] => 0
                                    [currency_code] => EUR
                                    [currency_symbol] => €
                                    [currency_minor_unit] => 2
                                    [currency_decimal_separator] => ,
                                    [currency_thousand_separator] => .
                                    [currency_prefix] => €
                                    [currency_suffix] =>
                                )

                            [catalog_visibility] => visible
                            [extensions] => Array
                                (
                                )

                        )

                )

            [items_count] => 1
            [items_weight] => 0
            [cross_sells] => Array
                (
                )

            [needs_payment] =>
            [needs_shipping] => 1
            [has_calculated_shipping] => 1
            [fees] => Array
                (
                )

            [totals] => Array
                (
                    [total_items] => 1000
                    [total_items_tax] => 0
                    [total_fees] => 0
                    [total_fees_tax] => 0
                    [total_discount] => 0
                    [total_discount_tax] => 0
                    [total_shipping] => 0
                    [total_shipping_tax] => 0
                    [total_price] => 0
                    [total_tax] => 0
                    [tax_lines] => Array
                        (
                        )

                    [currency_code] => EUR
                    [currency_symbol] => €
                    [currency_minor_unit] => 2
                    [currency_decimal_separator] => ,
                    [currency_thousand_separator] => .
                    [currency_prefix] => €
                    [currency_suffix] =>
                )

            [errors] => Array
                (
                )

            [payment_requirements] => Array
                (
                    [0] => products
                )

            [extensions] => Array
                (
                    [woocommerce-gift-cards] => Array
                        (
                            [balance] => 587005
                            [is_using_balance] => 1
                            [pending_total] => 0
                            [cart_total] => 1000
                            [available_total] => 1000
                            [account_giftcards] => Array
                                (
                                    [0] => Array
                                        (
                                            [code] => XXXX-XXXX-XXXX-NYHX
                                            [amount] => 1000
                                        )

                                )

                            [applied_giftcards] => Array
                                (
                                )

                        )

                )

        )

    [headers] => Array
        (
            [Nonce] => 832c7a90d8
            [Nonce-Timestamp] => 1664808868
            [User-ID] => 1
            [X-WC-Store-API-Nonce] => 832c7a90d8
        )

    [status] => 200
    [links:protected] => Array
        (
        )

    [matched_route:protected] =>
    [matched_handler:protected] =>
)
```
</details>

**Comment by wavvves** [comment]
IMO this should not be a high priority, as it affects third parties using our endpoint at their responsibility of parsing results while on PHP >=8.0

We do have a problem with consistency since we're returning different datatypes (array vs objects), I tried to reproduce this quickly but couldn't on the extensions endpoint.

I think this should be closed and linked to a new issue that has a more clear description of the problem and what to do on our side.

We also bring it to a discussion as starting using the correct data types alone will probably surface retro compatibility problems.

As soon as I get a couple of minutes to spare, I'll open and link a new issue.

**Comment by wavvves** [comment]
@PanosSynetos to not get blocked by this, and have a simple solution I recommend you typecast it to object.

eg.
`$item_data['quantity_limits'] = (object) $item_data['quantity_limits'] `

It will still work no matter how we decide to fix this.

**Comment by PanosSynetos** [comment]
Thanks Paulo. Yes, that's what I plan to do in our plugins.

Let me know if you need help reproducing this, I can reproduce it consistently.

**Comment by mikejolley** [comment]
`/wc/store/v1/cart/extensions` makes use of `/wc/store/v1/cart` directly:

```
return rest_ensure_response( wc()->api->get_endpoint_data( '/wc/store/v1/cart' ) );
```

I presume the issue arises from this re-use. The data should be of `object` type. Looking at `get_endpoint_data` from woo core, this makes use of both `json_encode` and `json_decode`, so this is likely where the object is being converted into an array. Perhaps we can move away from this function to resolve.

**Comment by PanosSynetos** [comment]
I wanted to mention, there are more properties that their type changes, not inside the `items` key but top level too.

```php
$data = $response->get_data();
```

![image](https://user-images.githubusercontent.com/2484390/196641380-7e7ec068-c8a3-4b13-b568-b052a2a9faaf.png)
